src/transformers/models/bloom/modeling_bloom.py

```diff
@@ -0,0 +1,990 @@
+    Pre-process the alibi tensor for padding.
+
+    Args:
+        alibi: ([`torch.tensor`], *required*):
+            alibi tensor to pre-process
+        attention_mask: ([`torch.tensor`], *required*):
+            attention mask to pre-process"""
+    # Remove the padding.
+    index_x, index_y = torch.where(attention_mask == 0.0)
+    batches_to_modify = torch.unique(index_x)
+
+    new_alibi = torch.clone(alibi.detach())
+    for i, index in enumerate(batches_to_modify):
+        slice_to_modify = torch.clone(new_alibi[index * num_heads : (index + 1) * num_heads].detach())
+
+        begin, end = torch.where(index_x == index)[0][0], torch.where(index_x == index)[0][-1]
+        if begin != end:
+            shifted_slice = slice_to_modify[:, :, : -(end - begin + 1)]
+
+            begin_position_to_extract, end_position_to_extract = index_y[begin], index_y[end] + 1
+
+            new_alibi[index * num_heads : (index + 1) * num_heads][:, :, begin_position_to_extract:] = 0.0
+            new_alibi[index * num_heads : (index + 1) * num_heads][:, :, end_position_to_extract:] = shifted_slice[
+                :, :, begin_position_to_extract:
+            ]
+
+    return new_alibi
+
+
+def split_tensor_along_last_dim(tensor, num_partitions, contiguous_split_chunks=False):
+    """Split a tensor along its last dimension.
+
+    Args:
+        tensor: ([`torch.tensor`], *required*):
+            input tensor to split
+        num_partitions ([`int`], *required*):
+            number of partitions to split the tensor
+        contiguous_split_chunks ([`bool`], *optional*, default=`False`)::
+            If True, make each chunk contiguous in memory.
+    """
+    # Get the size and dimension.
+    last_dim = tensor.dim() - 1
+    numerator, denominator = tensor.size()[last_dim], num_partitions
+    if not (numerator % denominator == 0):
+        raise ValueError(f"{numerator} is not divisible by {denominator}")
+    last_dim_size = numerator // denominator
+    # Split.
+    tensor_list = torch.split(tensor, last_dim_size, dim=last_dim)
+    # Note: torch.split does not create contiguous tensors by default.
+    if contiguous_split_chunks:
+        return tuple(chunk.contiguous() for chunk in tensor_list)
+
+    return tensor_list
+
+
+def attention_mask_func(attention_scores, attention_mask, causal_mask):
+    if attention_mask.dtype == torch.bool:
+        attention_mask_bool = ~attention_mask
+    else:
+        attention_mask_bool = (1 - attention_mask).bool()
+
+    query_length, key_length, n_heads = attention_scores.size(2), attention_scores.size(3), attention_scores.size(1)
+    padded_causal_mask = (
+        attention_mask_bool[:, None, key_length - query_length : key_length, None]
+        + ~causal_mask[:, :, key_length - query_length : key_length, :key_length]
+    ).bool()
+    index_x, index_y = torch.where(attention_mask == 0)
+    for i, index in enumerate(index_x):
+        padded_causal_mask[index, :, :, index_y[i]] = True
```

Comment:
Is this actually needed? From what I understand you're trying to `pad` the keys. This should be doable by a simple `add` when you construct the the `padded_causal_mask`

Something like (not tested):
```python
    padded_causal_mask = (
        attention_mask_bool[:, None, key_length - query_length : key_length, None]
        + attention_mask_bool[:, None, None, :key_length]
        + ~causal_mask[:, :, key_length - query_length : key_length, :key_length]
    ).bool()
```

Reply:
Should be fixed on : 7db6afe808e41e66ea20ff56db6f675fead03803

Reply:
Seems to be fixed, thanks for the correction! Fixed it here: 1450762f1584b549840abd622e3582885cb6fe90

Reply:
I'm a bit surprised you need this in fact. As you shouldn't care about running things to padded_tokens ...